src/nav/AppNavigator.js

```diff
@@ -82,9 +89,32 @@ export type AppNavigatorParamList = {|
   +'selectable-options': RouteParamsOf<typeof SelectableOptionsScreen>,
 |};
 
+/**
+ * A common supertype for navigation props in this and descendant navigators.
+ *
+ * In particular this contains methods like `push`, with all the main
+ * navigator's routes available, and `addListener` with the events that
+ * aren't specific to the stack navigator.
+ */
+export type AppNavigationMethods = StackActionHelpers<AppNavigatorParamList> &
+  NavigationHelpers<AppNavigatorParamList> &
+  EventConsumer<EventMapCore<NavigationState<AppNavigatorParamList>>>;
+
+/**
+ * The type of the `navigation` prop for screens in this navigator.
+ */
 export type AppNavigationProp<
   +RouteName: $Keys<AppNavigatorParamList> = $Keys<AppNavigatorParamList>,
-> = StackNavigationProp<GlobalParamList, RouteName>;
+> = StackNavigationProp<AppNavigatorParamList, RouteName> &
+  // Intersecting with the methods type should be redundant -- the
+  // StackNavigationProp type should already be a subtype of it.  But when
+  // we check that below, we'd get puzzling errors.  So add it explicitly.
+  AppNavigationMethods;
```

Comment:
> * The relevant class, I think, is `PrivateValueStore`. That's the only non-React-component `class` in `types/@react-navigation`. On one of its properties, there's a comment saying "UGLY HACK! DO NOT USE THE TYPE!!!"

Yeah. I think that whole `PrivateValueStore` type is there to do something that (a) doesn't describe anything about the actual runtime values -- there's no actual object floating around with properties `a` and `b` and `c` -- and (b) is there in order to be able to use a particular TypeScript feature. Namely, in the definition of `CompositeNavigationProp`, in the "conditional types" that are quoted commented out because TsFlower doesn't support them, you can see things like this:
```
/**
 * Event consumer config should refer to the config specified in the first type
 * This allows typechecking `addListener`/`removeListener`
 */
A extends NavigationProp<any, any, any, any, infer E> ? E : {}
```
What that's doing is taking a type `A` (a type parameter of `CompositeNavigationProp`), which is expected to be some `NavigationProp<…>` type, and extracting out from it what its `EventMap` type should be, i.e. the fifth type parameter of `NavigationProp`.

And then what that comment in `PrivateValueStore` is saying:
```
    /**
     * UGLY HACK! DO NOT USE THE TYPE!!!
     *
     * TypeScript requires a type to be used to be able to infer it.
     * The type should exist as its own without any operations such as union.
     * So we need to figure out a way to store this type in a property.
```
is that that `infer E` is only going to work if the thing being inferred appears in a nice direct way in the definition of the `NavigationProp` type. If you had e.g.
```ts
type Union<T, U> = T | U;
```
then saying `A extends Union<any, infer U> ? U : {}` couldn't usefully work, because for any given type A there may be all kinds of different types U that would fit. I'm not sure what TS's exact rules are here, but apparently they don't allow inferring the `EventMap` type from the way it's used for real in `NavigationProp`.

So the react-navigation authors' solution has been to introduce this `PrivateValueStore` type, and end the `NavigationProp` definition with `& PrivateValueStore<ParamList, RouteName, EventMap>`. Then
```ts
A extends NavigationProp<any, any, any, any, infer E> ? E : {}
```
becomes basically just a shorthand for saying
```ts
A extends NavigationProp<any, any, any, any, any> ? A['']['c'] : {}
```
with no `infer`.

Anyway, so that's what I've worked out `PrivateValueStore` is doing. Given that, I think for any type errors that involve it, we shouldn't worry about them potentially being signs of a real problem.

Reply:
> I haven't yet quite understood why we give `AppNavigationMethods` the non-stack-nav-specific `addListener`, while we _do_ give it the stack-nav-specific method `push` (as mentioned in `AppNavigationMethods`'s jsdoc).

Basically, I know that the `push` method is available when you're on a screen inside a child navigator (we've been using the functionality since forever via `NavigationService.dispatch`; and empirically it is there, and works, as a method on the navigation prop too, as seen by manually testing the tip of this branch); but I'm not sure whether events like `gestureCancel` are. I don't have as clear an idea of what those are meant to do, to understand whether I'd think they *should* be available there, either.

I did find one indication that they're not supposed to be available -- it's that `CompositeNavigationProp` type I discussed above in a different context. Here's the definition, from `node_modules/@react-navigation/core/src/types.tsx`:
```ts
export type CompositeNavigationProp<
  A extends NavigationProp<ParamListBase, string, any, any>,
  B extends NavigationHelpersCommon<ParamListBase, any>
> = Omit<A & B, keyof NavigationProp<any>> &
  NavigationProp<
    /**
     * Param list from both navigation objects needs to be combined
     * For example, we should be able to navigate to screens in both A and B
     */
    (A extends NavigationHelpersCommon<infer T> ? T : never) &
      (B extends NavigationHelpersCommon<infer U> ? U : never),
    // …
    /**
     * Event consumer config should refer to the config specified in the first type
     * This allows typechecking `addListener`/`removeListener`
     */
    A extends NavigationProp<any, any, any, any, infer E> ? E : {}
  >;
```

And here's the docs on what that `CompositeNavigationProp` type is supposed to mean:
https://reactnavigation.org/docs/typescript/#combining-navigation-props

So that definition takes the trouble to ensure that the routes from both parent and child are available; but for the events, it only includes those from the child.

Reply:
> When I imagine why that might be, I think vaguely of function subtyping, and things being in "input" and "output" positions. Am I onto something there?

And yeah, I think that's the kind of issue that causes this to be a type error. Specifically:
* The `AppNavigationProp` types have a larger set of events supported by their `addListener` method.
* The `AppNavigationMethods` type offers a smaller set of events.
* That means that an `AppNavigationMethods` value truly shouldn't be usable as an `AppNavigationProp` value. But ideally it shouldn't be an obstacle to taking an `AppNavigationProp` value and using it as an `AppNavigationMethods` -- the prop type should be a subtype of the methods type.
* That ideal would be met if all the relevant pieces of each of these types had the right things covariant and contravariant, aka had the right things marked as "output positions" and "input positions". Oh also: and had inexact object types in the necessary places, so that a type with more properties is a subtype of one with a subset of them.
* But apparently that's not the case.

Reply:
Thanks for that explanation!